File: old/algo/17.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        def dfs(index: int, path: str):
            if len(path) == len(digits):
                result.append(path)
                return
            
            for i in range(index, len(digits)):
                for j in dic[digits[i]]:
                    dfs(i + 1, path + j)

        dic = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
        }

        result = []

        if not digits:
            return []
        
        dfs(0, "")
        
        return result

digits = "23"

s = Solution()
print(s.letterCombinations(digits))

File: algo/17.kt
class Solution {
    fun letterCombinations(digits: String): List<String> {
        if (digits.isEmpty()) {
            return listOf()
        }
        val result = mutableListOf<String>()
        val numberToLetters = mapOf(
            '2' to "abc",
            '3' to "def",
            '4' to "ghi",
            '5' to "jkl",
            '6' to "mno",
            '7' to "pqrs",
            '8' to "tvu",
            '9' to "wxyz"
        )

        fun visit(start: Int, path: String) {
            if (digits.length == path.length) {
                result.add(path)
                return
            }

            for (i in start..digits.lastIndex) {
                for (ch in numberToLetters.getValue(digits[i])) {
                    visit(i + 1, path + ch)
                }
            }
        }

        visit(0, "")
        return result
    }
}

//val digits = "23"
val digits = ""
//val digits = "2"
println(Solution().letterCombinations(digits))
